# Лабораторна робота №1
## Завдання 2.1
### Виконав Назаренко Василь, 5 курс, "статистика"
#### Варіант №7

#####                                УМОВА

Розглянемо модель вигляду
$$ {X}_n = {x}_0 + {\theta{b}_n}+{\sigma}{R}_n $$,
у якій $ {R}_n = {\sum_{i=1}^{n}}{\varepsilon}_i, де{ \varepsilon}_i {\sim} = \mathcal{N}(0,1), {i \geq 0}, незалежні. $

Написати програму для обчислення:

а) оцінки параметра ${\theta } $ за останнім спостереженням;

б) ОМНК параметра $ {\theta } $  при відомому $ {x}_0$;

в) ОМНК параметрів ${x}_0$ та ${\theta}$;

г) оцінки параметра ${\sigma}$(при відомому ${\theta}$).

Згенерувати 1000 реалізацій моделі та знайти середні та дисперсії оцінок для n=50,100,200,500,1000. Порівняти якість різних оцінок параметра ${\theta}$. Чи підтверджують одержані результати конзистентність оцінок? Виконати те саме завдання для випадку, коли величини {${\varepsilon}_i,{i \geq 0}$} - незалежні та рівномірно розподілені на відрізку [-1,1].

Значення параметрів для моделі:

7) ${b}_n = {(8{n}^3-3n)}^{1/3}, {x}_0 = 4, {\theta} = 0.1, {\sigma} = 2$

In [257]:
import numpy as np

n = 10000
ksi_norm = np.random.normal(0, 1, n)
ksi_uniform =np.random.uniform(-1, 1, n)
X_0 = 4
theta = 0.1
sigma = 2
R_norm = [np.sum(ksi_norm[0:i]) for i in range(1,n+1)]
R_uniform = [np.sum(ksi_uniform[0:i]) for i in range(1,n+1)]

sizes = [50, 100, 200, 500, 1000, 10000]

In [258]:
def b(n):
    if n==0:
        return 0
    else:
        return (8*(n)**3-3*n)**(1/3)
#b(1) - b(0)

In [259]:
X_norm = [X_0 + theta*b(i)+sigma*R_norm[i-1] for i in range(1, n+1)]
X_uniform = [X_0 + theta*b(i)+sigma*R_uniform[i-1] for i in range(1, n+1)]

### а)
оцінка $ \theta $ за останнім спостереженням

$ \theta = \frac{{X}_n - {X}_0}{{b}_n}$

In [260]:
def theta_hat1(n, X):
    return (X[n-1]-X_0)/(b(n))

In [261]:
print("theta_hat1:")
print("{:6s} {:12s} {:12s} {:12s}  {:12s}".format("size", "norm Mean", "norm Var", "uniform Mean", "uniform Var"))
for size in sizes:
    list_norm = [theta_hat1(i,X_norm) for i in range(1,size+1)]
    list_uniform = [theta_hat1(i,X_uniform) for i in range(1,size+1)]
    
    print("{:4d}   {:.6f}     {:.6f}     {:.6f}      {:.6f}"
              .format(size, np.mean(list_norm), np.var(list_norm), np.mean(list_uniform), np.var(list_uniform)))

theta_hat1:
size   norm Mean    norm Var     uniform Mean  uniform Var 
  50   0.166325     0.028364     0.160400      0.007486
 100   0.186528     0.015076     0.094795      0.008247
 200   0.161346     0.008318     0.087326      0.004277
 500   0.158919     0.003382     0.078856      0.001780
1000   0.151740     0.001761     0.078709      0.000943
10000   0.117802     0.000322     0.096790      0.000169


### б)
ОМНК параметра $ {\theta } $  при відомому $ {x}_0$;

${\hat{\theta}_n } = \frac{\sum_{i=1}^{n}\Delta{X}_i\Delta{b}_i}{\sum_{i=1}^{n}(\Delta{b}_i)^2}$

In [262]:
def theta_hat2(n, X):
    return np.sum([(X[i+1] - X[i])*(b(i+1) - b(i)) for i in range(0,n)])\
    /np.sum([(b(i+1) - b(i))**2 for i in range(0, n)])

In [263]:
print("theta_hat2:")
print("{:6s} {:12s} {:12s} {:12s}  {:12s}".format("size", "norm Mean", "norm Var", "uniform Mean", "uniform Var"))
for size in sizes:
    list_norm = [theta_hat2(i,X_norm) for i in range(1, size)]
    list_uniform = [theta_hat2(i,X_uniform) for i in range(1, size)]
    print("{:4d}   {:.6f}     {:.6f}     {:.6f}      {:.6f}"
          .format(size, np.mean(list_norm), np.var(list_norm), np.mean(list_uniform), np.var(list_uniform)))

theta_hat2:
size   norm Mean    norm Var     uniform Mean  uniform Var 
  50   0.094276     0.035258     0.164421      0.023188
 100   0.145660     0.020489     0.094230      0.016524
 200   0.137949     0.010397     0.086245      0.008387
 500   0.148071     0.004269     0.078000      0.003411
1000   0.145736     0.002156     0.078129      0.001758
10000   0.117000     0.000325     0.096687      0.000253


### в)
в) ОМНК параметрів ${x}_0$ та ${\theta}$;

$ {\hat{X}_0} = {X}_1 - {\hat{\theta}_n} {b}_1$

${\hat{\theta}_n } = \frac{\sum_{i=2}^{n}\Delta{X}_i\Delta{b}_i}{\sum_{i=2}^{n}(\Delta{b}_i)^2}$

In [264]:
def theta_hat(n,X):
    return np.sum([(X[i+1] - X[i])*(b(i+1) - b(i)) for i in range(1,n)])\
    /np.sum([(b(i+1) - b(i))**2 for i in range(1, n)])
def X_0(n,X):
    return  X[0] - theta_hat(n,X)*b(1)

#X0(1000, X_uniform)

In [265]:
print("theta_hat:")
print("{:6s} {:12s} {:12s} {:12s}  {:12s}".format("size", "norm Mean", "norm Var", "uniform Mean", "uniform Var"))
for size in sizes:
    list_norm = [theta_hat(i,X_norm) for i in range(2, size)]
    list_uniform = [theta_hat(i,X_uniform) for i in range(2, size)]
    print("{:4d}   {:.6f}     {:.6f}     {:.6f}      {:.6f}"
          .format(size, np.mean(list_norm), np.var(list_norm), np.mean(list_uniform), np.var(list_uniform)))

theta_hat:
size   norm Mean    norm Var     uniform Mean  uniform Var 
  50   0.010055     0.067466     0.084619      0.009603
 100   0.100874     0.041388     0.048653      0.006212
 200   0.113570     0.020789     0.060953      0.003341
 500   0.137293     0.008695     0.066516      0.001367
1000   0.139934     0.004364     0.071867      0.000767
10000   0.116273     0.000515     0.095891      0.000181


In [267]:
print("X_0:")
print("{:6s} {:12s} {:12s} {:12s}  {:12s}".format("size", "norm mean X0", "norm Var X0", "uniform Mean X0", "uniform Var X0"))
for size in sizes:
    list_norm = [X_0(i,X_norm)  for i in range(2, size)]
    list_uniform = [X_0(i,X_uniform)  for i in range(2, size)]
    print("{:4d}   {:.6f}     {:.6f}     {:.6f}         {:.6f}"
          .format(size, np.mean(list_norm), np.var(list_norm), np.mean(list_uniform), np.var(list_uniform)))

X_0:
size   norm mean X0 norm Var X0  uniform Mean X0  uniform Var X0
  50   5.485826     0.197271     4.117832         0.028079
 100   5.330528     0.121020     4.179333         0.018164
 200   5.308818     0.060789     4.158301         0.009770
 500   5.268252     0.025425     4.148789         0.003997
1000   5.263737     0.012761     4.139637         0.002243
10000   5.304195     0.001506     4.098557         0.000530


### г)
$ ({\hat{\sigma}})_n^2 = \frac{\sum_{i=1}^{n}(\Delta{X}_i)^2}{n}$

In [268]:
def sigma2_hat(n,X):
    return np.sum([(X[i]-X[i-1])**2 for i in range(1, n)])/n

In [270]:
print("sigma2_hat:")
print("{:6s} {:12s} {:12s} {:12s}  {:12s}".format("size", "norm Mean", "norm Var", "uniform Mean", "uniform Var"))
for size in sizes:
    list_norm = [sigma2_hat(i,X_norm) for i in range(1, size)]
    list_uniform = [sigma2_hat(i,X_uniform) for i in range(1, size)]
    print("{:4d}   {:.6f}     {:.6f}     {:.6f}      {:.6f}"
          .format(size, np.mean(list_norm), np.var(list_norm), np.mean(list_uniform), np.var(list_uniform)))

sigma2_hat:
size   norm Mean    norm Var     uniform Mean  uniform Var 
  50   3.950244     0.699415     1.135301      0.051207
 100   3.707427     0.427712     1.199365      0.030644
 200   3.583443     0.241741     1.301094      0.025761
 500   3.700243     0.108269     1.327216      0.010920
1000   3.800292     0.064561     1.349201      0.006027
10000   4.003079     0.013911     1.365054      0.000709


# Висновок
Видно, що при збільшенні $ n $ від 50 до 1000 дав не найкращий результат, який хотілось би. Але вже при $ n $ = 10000, стає більш менш зрозуміло, що при збільшенні $ n $, можна сказати, що оцінка є конзистентною. 

Також, хочу добавити, що на результати може вплинути, як саме початково згенерується випадкова вибірка. Бувало таке, що оцінка $\theta$ при збільшенні $ n $ вже відразу коливалася навколо $ 0.1 $ ( тобто, мого початкового значення ). Не став повторювати код ще раз, для більш кращого результату, оскільки при n = 10000 компютер вже доволі таки довго рахує це.

Також, не зовсім зрозуміло, що має бути з оцінкой $ {X}_0 $, якщо вона повинна прямувати до мого заданого значення, то тут точно сказати важко. Хіба що, при рівномірному розподілі, вона більш менш прямує до 4. Що не можно сказати про нормальний розподіл(там при різних запусках програми, вона то в окулі одиниці, то двійки і т.д.). В даному випадку, $ {X}_0 $ біля 5.3 при нормальному розподілі та біля 4.1 при рівномірному.